### PR TITLE
WT-12278 walk tree statistics should consider the impact of urgent queue

### PR DESCRIPTION
**Pull request by y123456yz**
evict walk tree bug fix, walk tree stastic should consider the impact of urgent queue.

ignoring urgent queue's influence will have some impact on statistics and performance.

### CODE REVIEW

**Comment by y123456yz** [comment]
WT-12278

**Comment by luke-pearson** [comment]
merge develop in please

File: src/evict/evict_lru.c [hunk 1]
```diff
@@ -2091,6 +2099,8 @@ __evict_walk_tree(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, u_int max_ent
         WT_ASSERT(session, evict->ref == NULL);
         if (!__evict_push_candidate(session, queue, evict, ref))
             continue;
+
+wark_statistical:
```
Comment (quchenhao) [inline src/evict/evict_lru.c:2103]: typo

File: src/evict/evict_lru.c [hunk 2]
```diff
@@ -2099,8 +2109,14 @@ __evict_walk_tree(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, u_int max_ent
         if (F_ISSET(ref, WT_REF_FLAG_INTERNAL))
             internal_pages_queued++;
 
-        __wt_verbose(session, WT_VERB_EVICTSERVER, "select: %p, size %" WT_SIZET_FMT, (void *)page,
-          page->memory_footprint);
+        if (urgent_queued_flag)
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/evict/evict_lru.c:2169]: @y123456yz Have you thought about this?

File: src/evict/evict_lru.c [hunk 3]
```diff
@@ -2156,11 +2175,28 @@ __evict_walk_tree(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, u_int max_ent
         if (F_ISSET(ref, WT_REF_FLAG_INTERNAL))
             internal_pages_queued++;
 
-        __wt_verbose(session, WT_VERB_EVICTSERVER, "walk select: %p, size %" WT_SIZET_FMT,
-          (void *)page, page->memory_footprint);
+        if (WT_VERBOSE_LEVEL_ISSET(session, WT_VERB_EVICTSERVER, WT_VERBOSE_DEBUG_2)) {
+            if (urgent_queued_flag)
```
Comment (quchenhao) [inline src/evict/evict_lru.c:2179]: I was thinking that we only need to aggregate the number of pages and sizes. We don't need to print out the address of pages. Do you find the address useful for debugging?

**Comment by quchenhao** [comment]
I still don't know how valuable this change is for debugging. Adding @Siddhartha8899 @kommiharibabu to have a look.

**Comment by Siddhartha8899** [comment]
**The modifications in this PR are divided into three primary areas:**

1. Count root pages as internal pages for stat purposes.
2. Include urgently queued pages in the overall queue count and ensure accurate reflection in the stats.
3. Print the pointer and memory footprint for each queued page.

**Responses to the Changes:**

- (**1**): Agreeing with the Decision
I agree with the inclusion of root pages as internal pages makes sense.

- (**3**): Disagreement Due to Excess Information
Printing the pointer for each page is excessive. Suggesting instead that a simple count of queued pages is more useful and relevant, which already exists today.

-  (**2**) In-depth Discussion:

    -   **Background and Complexity**
          - Originally, the system was designed with a single queue. Pages were added incrementally, and progress was tracked until reaching a target page count.
          - The addition of an urgent queue complicates the existing logic, particularly the "give up" logic, which may indicate that no pages have been added if only the urgent queue is used.
  
   - **Proposed Solution**
      Suggesting the separation of logic into two parts:
        - The first part monitors the ordinary queue and tracks pages added to it.
        - The second part evaluates all pages in both queues to make decisions about abandoning the process or continuing with the b-tree walk.

File: src/evict/evict_lru.c [hunk 4]
```diff
@@ -1967,7 +1970,8 @@ __evict_walk_tree(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, u_int max_ent
          */
         give_up = !__wt_cache_aggressive(session) && !WT_IS_HS(btree->dhandle) &&
           pages_seen > min_pages &&
-          (pages_queued == 0 || (pages_seen / pages_queued) > (min_pages / target_pages));
+          (pages_queued + pages_urgent_queued == 0 ||
```
Comment (quchenhao) [inline src/evict/evict_lru.c:1973]: This is not only a stat change. What's the intention behind this particular change?

**Review by quchenhao — APPROVED** [review]
I think I understand this change and it now makes sense to me.

lgtm

**Comment by quchenhao** [comment]
@Siddhartha8899 Please have another look.

**Comment by Siddhartha8899** [comment]
I am somewhat hesitant about moving forward with this change, as the logic was initially intended for the ordinary queue. Unfortunately the changes would need to be more complex and I fail to see the trade-off justifying the effort.

Explaining further complexities, that may not have been considered.

- Initially, the code was designed specifically for ordinary queues. In the walk process, we select an ordinary queue, identify a relevant tree, and determine target pages based on the tree's contributions to cache and the available slots in the ordinary queue. ([here](https://github.com/y123456yz/wiredtiger/blob/3a432ff8e2b5783743b351d40a1cba159b037045/src/evict/evict_lru.c#L1794))
- Then, the [process](https://github.com/y123456yz/wiredtiger/blob/3a432ff8e2b5783743b351d40a1cba159b037045/src/evict/evict_lru.c#L2048) involves walking the tree and selecting pages that align with the server's strategy (dirty, clean, update). The goal is to chose pages that meet these criteria. Failure to queue pages would result in [deprioritising](https://github.com/y123456yz/wiredtiger/blob/3a432ff8e2b5783743b351d40a1cba159b037045/src/evict/evict_lru.c#L2122) the tree or [walking](https://github.com/y123456yz/wiredtiger/blob/3a432ff8e2b5783743b351d40a1cba159b037045/src/evict/evict_lru.c#L1928) it differently. For urgent eviction, there is no check to ensure if the page meets the criteria.
- This process happens until the ordinary queue is filled ([here](https://github.com/y123456yz/wiredtiger/blob/3a432ff8e2b5783743b351d40a1cba159b037045/src/evict/evict_lru.c#L1917)) based on the target pages.
 
The above clearly indicates that the original design and decision-making was tailored to one queue and I don't see any valuable trade-off to modify the code change that justifies the effort.

**Comment by y123456yz** [comment]
> I am somewhat hesitant about moving forward with this change, as the logic was initially intended for the ordinary queue. Unfortunately the changes would need to be more complex and I fail to see the trade-off justifying the effort.
> 
> Explaining further complexities, that may not have been considered.
> 
> * Initially, the code was designed specifically for ordinary queues. In the walk process, we select an ordinary queue, identify a relevant tree, and determine target pages based on the tree's contributions to cache and the available slots in the ordinary queue. ([here](https://github.com/y123456yz/wiredtiger/blob/3a432ff8e2b5783743b351d40a1cba159b037045/src/evict/evict_lru.c#L1794))
> * Then, the [process](https://github.com/y123456yz/wiredtiger/blob/3a432ff8e2b5783743b351d40a1cba159b037045/src/evict/evict_lru.c#L2048) involves walking the tree and selecting pages that align with the server's strategy (dirty, clean, update). The goal is to chose pages that meet these criteria. Failure to queue pages would result in [deprioritising](https://github.com/y123456yz/wiredtiger/blob/3a432ff8e2b5783743b351d40a1cba159b037045/src/evict/evict_lru.c#L2122) the tree or [walking](https://github.com/y123456yz/wiredtiger/blob/3a432ff8e2b5783743b351d40a1cba159b037045/src/evict/evict_lru.c#L1928) it differently. For urgent eviction, there is no check to ensure if the page meets the criteria.
> * This process happens until the ordinary queue is filled ([here](https://github.com/y123456yz/wiredtiger/blob/3a432ff8e2b5783743b351d40a1cba159b037045/src/evict/evict_lru.c#L1917)) based on the target pages.
> 
> The above clearly indicates that the original design and decision-making was tailored to one queue and I don't see any valuable trade-off to modify the code change that justifies the effort.


I made a change to this PR later, which does not affect the number of pages of the ordinary queue, but only affects give_up.
https://github.com/y123456yz/wiredtiger/blob/develop__evict_walk_tree_bug_fix/src/evict/evict_lru.c#L2164-L2174

The only things that affect the selection of candidate pages in a normal queue are as follows:
https://github.com/y123456yz/wiredtiger/blob/develop__evict_walk_tree_bug_fix/src/evict/evict_lru.c#L1973-L1974

If the logic of L1973-L1974 does not require it, here I think pages_seen should only count the ordinary queue, not the urgent queue page.

Here you confirm, I will make the modification according to your idea，thanks.

**Comment by Siddhartha8899** [comment]
The reason I think it should not be added to the give-up logic is that this logic is primarily designed for situations where we cannot queue pages that don't comply with the server's strategy. In such cases, the server might de-prioritise the tree or adopt a different walking approach to ensure it finds the appropriate pages to evict. 

Given the current modifications, since urgent eviction does not adhere to the eviction strategy, it seems incorrect not to give up. As I mentioned, implementing this would require complex/significant changes. To me, the benefits do not justify all the effort involved.

Pages seen include both ordinary and urgent, that's correct. However, we do maintain separate stats for what's queued in each queue - `pages queued for eviction`, and `pages queued for urgent eviction during the walk`.

**Comment by y123456yz** [comment]
@Siddhartha8899  
Following your instructions, I have change the code, please review again.
thanks.

File: src/evict/evict_lru.c [hunk 5]
```diff
@@ -2148,24 +2159,33 @@ __evict_walk_tree(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, u_int max_ent
...
 
+walk_statistical:
+        if (!urgent_queued_flag) {
+            ++evict;
+            ++btree->evict_walk_progress;
+            ++pages_queued;
+            pages_queued_memory += page->memory_footprint;
+        } else {
+            ++pages_urgent_queued;
+            pages_urgent_queued_memory += page->memory_footprint;
```
Comment (Siddhartha8899) [inline src/evict/evict_lru.c:2171]: Thanks @y123456yz, can you help me understand the benefits of having the average page memory footprint for each incremental walk? Personally, I have never had the need for this information.

**Comment by kommiharibabu** [comment]
The last three months have seen no development on this PR. I'm going to close it now. If the corrections and comments have been addressed, reopen it. Thanks.

**Pull request closed without merging**